scripts/bert/finetune_classifier.py

```diff
@@ -111,25 +144,29 @@ def evaluate():
         Ls.append(ls)
         step_loss += sum([L.asscalar() for L in Ls])
         metric.update([label], [out])
-    logging.info('validation accuracy: %s', metric.get()[1])
+    metric_nm, metric_val = metric.get()
+    if not isinstance(metric_nm, list):
+        metric_nm = [metric_nm,]
+        metric_val = [metric_val,]
+    metric_str = 'validation metrics:' + ','.join([i + ':{:.4f}' for i in metric_nm])
+    logging.info(metric_str.format(*metric_val))
 
 
-def train():
+def train(metric):
     """Training function."""
-    trainer = gluon.Trainer(model.collect_params(), args.optimizer,
-                            {'learning_rate': lr, 'epsilon': 1e-9})
+    trainer_w = gluon.Trainer(model.collect_params('.*weight'), args.optimizer, \
+                            {'learning_rate': lr, 'epsilon': 1e-9, 'wd':0.01})
+
+    # Do not apply weight decay on LayerNorm and bias terms
```

Comment:
This is very nice! Is there a reference for this idea? If you just found this to perform better than the previous code with single trainer, would you briefly comment what your observation was?

Reply:
That code aims for simplicity and readability, not for performance. I saw that kind of code on http://d2l.ai/chapter_deep-learning-basics/weight-decay.html#Weight-Decay-in-Gluon. :grin: